Feature: 어드민 특정 사장님 조회 API

songsunkook opened the pull request:
## 4번째 기록
---

**<변경 사항>**
Refactor: java/koreatech/in/repository/admin/AdminUserMapper > getUserById의 join문을 별도의 쿼리로 분리

## 3번째 기록
---

**<변경 사항>**
Refactor: Admin OwnerResponse 구조 개선
1. OwnerResponse가 UserResponse를 상속하도록 구조 변경
2. OwnerResponse에는 UserResponse 외에 사업자등록번호와 첨부파일 및 상점의 id들만을 가지도록 축소

**<변경 사유>**
1. OwnerResponse 개발 도중에 UserResponse가 생성되면서 코드 중복의 발생
    - UserResponse를 상속하도록 변경
    - Owner에서는 name 속성이 반드시 포함되어야 하므로 name에 required=true 속성을 추가하여 오버라이딩
2. OwnerResponse에 상점과 첨부파일 정보를 모두 담다보니 불필요하게 크기가 늘어남
    - 어드민 사장님 조회에 대한 새로운 POST, PUT 요청이 추가된다면 동일한 구조의 DTO로 만들어질 텐데 필요한 정보보다 불필요한 정보가 많이 포함됨

## 2번째 기록
---

**<변경 사항>**
1. 어드민 특정 사장님 조회용 API 추가 및 적용
4. 조회를 시도한 id의 신분이 사장님이 아닐 경우 인식 후 익셉션 Throw
5. Swagger 이슈로 동작하지 않던 일부 기존 메서드 버그 수정
    > AdminUserController.deleteUser
    > AdminUserController.undeleteUser

**<파일 변경 내용>**
1. Add
```
    - koreatech.in.dto.admin.user.response.OwnerResponse
    - koreatech.in.mapstruct.admin.user.OwnerConverter.toFullOwnerResponse(Owner owner)
    - koreatech.in.repository.admin.AdminUserMapper.getFullOwnerById(Integer id)
    - koreatech.in.resources.mapper.admin.AdminUserMapper <select id="getFullOwnerById">
    - koreatech.in.resources.mapper.admin.AdminUserMapper <resultMap id="FullOwnerResultMap">
```
2. Modify
```
    - koreatech.in.controller.admin.AdminUserController.getOwner()
    - koreatech.in.service.admin.AdminUserServiceImpl.getOwner(int ownerId)
    - koreatech.in.controller.admin.AdminUserController.deleteUser(Integer userId)
    - koreatech.in.controller.admin.AdminUserController.undeleteUser(Integer userId)
```

## 1번째 기록
---

Admin 페이지에서 id를 통해 사장님을 조회하면 기존에 존재하던 `OwnerMapper.getOwnerById()` 메서드를 통해 사장님 정보를 불러와 반환합니다.

**< 추가된 메서드 >**
controller/admin/AdminUserController
> ResponseEntity getOwner(int ownerId)

service/admin/AdminUserServiceImpl (AdminUserService 인터페이스에도 적용)
> OwnerResponse getOwner(int ownerId)

songsunkook committed:
feat: 어드민 특정 사장님 조회

Admin 페이지에서 id를 통해 사장님을 조회할 수 있는 Controller 및 Service 메서드 추가

Invidam commented on src/main/java/koreatech/in/controller/admin/AdminUserController.java:
```diff
@@ -198,4 +198,18 @@ ResponseEntity<NewOwnersResponse> getNewOwners(NewOwnersCondition condition) {
+    })
+    @RequestMapping(value = "/admin/users/owner/{ownerId}", method = RequestMethod.GET)
+    public @ResponseBody
+    ResponseEntity getOwner(@ApiParam(name = "ownerId", required = true) @PathVariable("ownerId") int ownerId) throws Exception {
```
```suggestion
    ResponseEntity getOwner(@ApiParam(name = "ownerId", required = true) @PathVariable("ownerId") int ownerId) {
```

필요치 않아 지울 수 있을 것 같아요~

Invidam replied:
반환 타입을 `ResponseEntity<OwnerResponse>` 변경해야 
swagger 상에 반환 타입이 표현되는 걸로 알고있어요~

현재)
![image](https://user-images.githubusercontent.com/71889359/231461779-b934f159-2236-4027-92f5-2c39c451ff77.png)

Invidam replied:
admin-user-student-get 쪽은 `id`를 사용했는데 여기는 `ownerId`을 사용하네요..

둘 중 하나는 바뀌어야 할 것 같습니다.
프론트쪽에서 무엇을 더 선호할지 물어보는 게 좋을 것 같아요~

daheeParkk commented on src/main/java/koreatech/in/service/admin/AdminUserService.java:
```diff
@@ -43,4 +44,6 @@ public interface AdminUserService {
 
     NewOwnersResponse getNewOwners(NewOwnersCondition condition);
+
+    OwnerResponse getOwner(int ownerId);
```
owner의 response DTO는 admin 패키지에서 새로 안 만들고 기존에 있던 거 써도 되나요?

Invidam commented on src/main/java/koreatech/in/service/admin/AdminUserServiceImpl.java:
```diff
@@ -379,6 +386,14 @@ public NewOwnersResponse getNewOwners(NewOwnersCondition condition) {
 
+    @Override
+    @Transactional(readOnly = true)
+    public OwnerResponse getOwner(int ownerId) {
+        Owner ownerInDB = Optional.ofNullable(ownerMapper.getOwnerById((long)ownerId))
+            .orElseThrow(() -> new BaseException(INQUIRED_USER_NOT_FOUND));
+        return OwnerConverter.INSTANCE.toOwnerResponse(ownerInDB);
+    }
+
```
**어드민**의 사장님 정보조회는
**일반 사장님**의 사장님 정보조회보다
많은 정보를 제공이 필요합니다!

따라서, (더 많은 정보 제공 위해) 별도의 DTO, Mapper가 필요할 것 같아요. 미리 말씀 못드렸네요.. 😅

Invidam reviewed (changes requested):
첫 작업인데도 컨벤션이 잘 지켜져있네요 👍

코드 리뷰 남겼으니 확인 부탁드려요😉~ 

기존 레거시 코드들이 많은데, 함께 고쳐나갑시다  \\^o^/

Invidam commented:
![image](https://user-images.githubusercontent.com/71889359/231468592-1cf467a6-8715-44f0-bdfa-d9e57f93709f.png)

참고해서 Git Conflict 수정 부탁드려요~ 

(모르는 부분 있으면 물어봐도 괜찮아요 ☺)

damiano1027 commented:
@Invidam @daheeParkk 
어제 선권이랑 얘기 잠깐 나누어본건데,

다희는 User 객체를 조회한다음, 해당 유저의 신원이 학생이 아니면 Exception을 throw하였고,
선권이는 Owner 객체로 바로 조회한다음, 따로 Exception을 throw하지 않았어요. (이미 owner로 조회하였기 때문에 신원이 사장님으로 보장되기 때문)

평소에는 생각없이 User 객체로만 조회했었는데,
선권이가 개발한 대로 처음부터 Owner 객체로 받아오는것도 불필요한 Exception 체크를 할 필요가 없어서 좋은 것 같은데,
어떻게 생각하시는지 궁금해요.

Kim-Juwon commented:
> @Invidam @daheeParkk 어제 선권이랑 얘기 잠깐 나누어본건데,
> 
> 다희는 User 객체를 조회한다음, 해당 유저의 신원이 학생이 아니면 Exception을 throw하였고, 선권이는 Owner 객체로 바로 조회한다음, 따로 Exception을 throw하지 않았어요. (이미 owner로 조회하였기 때문에 신원이 사장님으로 보장되기 때문)
> 
> 평소에는 생각없이 User 객체로만 조회했었는데, 선권이가 개발한 대로 처음부터 Owner 객체로 받아오는것도 불필요한 Exception 체크를 할 필요가 없어서 좋은 것 같은데, 어떻게 생각하시는지 궁금해요.

Invidam commented:
> @Invidam @daheeParkk 어제 선권이랑 얘기 잠깐 나누어본건데,
> 
> 다희는 User 객체를 조회한다음, 해당 유저의 신원이 학생이 아니면 Exception을 throw하였고, 선권이는 Owner 객체로 바로 조회한다음, 따로 Exception을 throw하지 않았어요. (이미 owner로 조회하였기 때문에 신원이 사장님으로 보장되기 때문)
> 
> 평소에는 생각없이 User 객체로만 조회했었는데, 선권이가 개발한 대로 처음부터 Owner 객체로 받아오는것도 불필요한 Exception 체크를 할 필요가 없어서 좋은 것 같은데, 어떻게 생각하시는지 궁금해요.

User이긴 하나, Owner이 아닌 회원(e.g. 학생)에 대해 `"조회한 회원이 존재하지 않습니다."` 라고 띄워주면 의미가 정확하지 않을 것 같아요!

songsunkook committed:
Merge branch 'develop' into feature/admin-get-owner-api

daheeParkk commented:
> User이긴 하나, Owner이 아닌 회원(e.g. 학생)에 대해 "조회한 회원이 존재하지 않습니다." 라고 띄워주면 의미가 정확하지 않을 것 같아요!

저는 이 경우에 대해 생각해 보지 못했는데, student의 id인 경우가 회원이 존재하지 않는 경우와 같은 에러 메세지가 발생하면 안될 것 같다는 느낌을 받았습니다 !!!

songsunkook committed:
feat: Owner DTO, Mapper 파일 생성

songsunkook committed:
refactor: AdminUserController import 컨벤션 적용

AdminUserController import문 순서 변경

songsunkook committed:
refactor: 어드민 특정 사장님 조회 API 개선

**<변경 전>**
기존에 존재하던 일반 유저용 사장님 조회 API를 가져다 사용
**<변경 후>**
어드민용 사장님 조회 API를 추가하여 사용

songsunkook committed:
refactor: NOT_OWNER 예외처리 추가

조회한 회원의 신원이 사장님이 아닐 경우 NOT_OWNER BaseException을 Throw하도록 수정

songsunkook committed:
fix: 기존 메서드 Swagger ApiParam 수정

파라미터가 JSON이 아닌 경우에 ApiParam > name을 기존 파라미터명이 아닌 부가설명을 붙여서 에러가 발생한 메서드를 수정했습니다.
**name -> value**

songsunkook committed:
Merge branch 'feature/admin-get-owner-api' of https://github.com/BCSDLab/KOIN_API into feature/admin-get-owner-api

# Conflicts:
#	src/main/java/koreatech/in/controller/admin/AdminUserController.java
#	src/main/java/koreatech/in/exception/ExceptionInformation.java
#	src/main/java/koreatech/in/service/admin/AdminUserService.java
#	src/main/java/koreatech/in/service/admin/AdminUserServiceImpl.java

Invidam commented on src/main/java/koreatech/in/dto/admin/user/response/OwnerResponse.java:
```diff
@@ -0,0 +1,168 @@
+    private String name;
+
+    @ApiModelProperty(notes = "휴대폰 번호", example = "010-0000-0000")
+    private String phoneNumber;
+
+    @ApiModelProperty(example = "STUDENT", required = true)
+    private UserType userType;
+
+    @ApiModelProperty(notes = "성별(남:0, 여:1)", example = "1")
+    private Integer gender;
+
+    @ApiModelProperty(notes = "인증 여부", example = "ture", required = true)
+    private Boolean isAuthed;
+
+    @ApiModelProperty(notes = "최근 로그인 날짜", example = "1680585195000")
+    private Date lastLoggedAt;
+
+    @ApiModelProperty(notes = "프로필 이미지 s3 URL", example = "static.koreatech.in/example.png")
+    private String profileImageUrl;
+
+    @ApiModelProperty(notes = "이메일 인증 토큰", example = "string")
+    private String authToken;
+
+    @ApiModelProperty(notes = "이메일 인증 토큰 만료 시간", example = "1680585195000")
+    private Date authExpiredAt;
+
+    @ApiModelProperty(notes = "비밀번호 초기화 토큰", example = "string")
+    private String resetToken;
+
+    @ApiModelProperty(notes = "비밀번호 초기화 토큰 만료 시간", example = "1680585195000")
+    private Date resetExpiredAt;
+
+    @ApiModelProperty(notes = "탈퇴 여부", example = "true")
+    private Boolean isDeleted;
+
+    @ApiModelProperty(notes = "회원가입 일자(생성 일자)", example = "1680585195000")
+    private Date createdAt;
+
+    @ApiModelProperty(notes = "업데이트 일자", example = "1680585195000")
+    private Date updatedAt;
```
UserResponse를 상속하여 구현하면 코드 중복을 줄일 수 있을 것 같아요

normal.UserResponse를 참고해보세요! 
특히 normal.UserResponse-name이 왜 있는지 고민해보세요

songsunkook replied:
name 속성이 OwnerResponse와 StudentResponse가 겹치지만 StudentResponse에서는 선택적이기 때문에 오버라이딩(?)하지 않았고, OwnerResponse에서는 필수적이기 때문에 required=true를 넣어 재정의했다고 생각됩니다.

UserResponse를 상속하면서 required 속성을 재지정하는 방향으로 진행해보겠습니다!

혹시 required를 정하는 명확한 기준이 있는지 궁금합니다. 저는 DB에서 NOT NULL 속성이 지정된 컬럼에 required=true를 넣어주는 것이 바람직해 보이는데 normal.OwnerResponse.name의 required는 DB 속성과는 연관이 없는 것 같습니다....

Invidam replied:
의도대로 잘 분석해줬네요!

NOT NULL이면 required = true 인건 맞지만,
Owner에서는 정책상 무조건 name이 있어서 required 옵션을 추가했어요.

Invidam commented on src/main/java/koreatech/in/dto/admin/user/response/OwnerResponse.java:
```diff
@@ -0,0 +1,168 @@
+    private String companyRegistrationCertificateImageUrl;
+*/
+
+    @ApiModelProperty(notes = "가게 목록")
+    @Valid
+    private List<Shop> shops;
+
+    @ApiModelProperty(notes = "첨부파일 목록")
+    @Valid
+    private List<Attachment> attachments;
+
+    @Getter
+    @Builder
+    @ApiModel("Admin_Shop_in_OwnerResponse")
```
어드민 특정사장님 조회 API에
상점 정보가 필요한가요?
상점 정보가 필요하다면 어느정도 범위까지 제공해야할까요?

songsunkook replied:
저는 어드민의 요청이기 때문에 가능한 모든 정보를 반환해야 한다고 생각했는데, 이렇게 되면 불필요하게 Response가 커질 수도 있을 것 같습니다. 기본적으로 어드민 사장님 조회 API에서는 해당 사장님의 상점 id와 상점명만 받아오고, 사장님이 소유한 상점 정보를 받아오는 별도의 어드민 API를 구현하는 것도 괜찮을 것 같습니다.

@Invidam  @Kim-Juwon @daheeParkk 
이 부분에 대해서 혹시 어떻게 생각하시는지 궁금합니다!

Invidam replied:
프론트쪽에 물어봐야 할 사항인 것 같아요~

Invidam commented on src/main/java/koreatech/in/exception/ExceptionInformation.java:
```diff
@@ -33,6 +34,7 @@ public enum ExceptionInformation {
     STUDENT_NUMBER_INVALID("학생의 학번 형식이 아닙니다.", 101015, HttpStatus.UNPROCESSABLE_ENTITY),
     STUDENT_MAJOR_INVALID("학생의 전공 형식이 아닙니다.", 101016, HttpStatus.UNPROCESSABLE_ENTITY),
     NOT_STUDENT("회원의 신원이 학생이 아닙니다.", 101017, HttpStatus.CONFLICT),
+    NOT_OWNER("회원의 신원이 사장님이 아닙니다.", 101018, HttpStatus.CONFLICT),
```
👍

Invidam commented on src/main/java/koreatech/in/service/admin/AdminUserServiceImpl.java:
```diff
@@ -396,6 +401,22 @@ public NewOwnersResponse getNewOwners(NewOwnersCondition condition) {
+            throw new BaseException(NOT_OWNER);
+        }
+
+        Owner fullOwnerInDB = Optional.ofNullable(adminUserMapper.getFullOwnerById(ownerId))
```
userInDB에도 사장님 정보가 있는데, 
fullOwnerInDB으로 사장님정보도 가져올 필요가 있을까요?
상점 or 첨부파일 정보만 가져오면 될 것 같아요!

songsunkook replied:
윗 주제와 충돌하는 내용이어서 더 고민해봐야 할 것 같습니다..

Invidam replied:
충돌하는 내용인 것도 맞지만,
userInDB를 가져올 때 사장님의 정보가 이미 들어와있는 상황인데, (사업자등록번호 등)
fullOwnerInDB를 가져올 때  (사업자등록번호 등의 )사장님 정보를 중복하여 가져오는 건 불필요해보여요!

Invidam reviewed (approved):
admin의 사장님 관련 로직들이 만들어졌네요 👍

리뷰 내용 확인부탁해요!
코드를 작성할 때, 꼭 필요한 기능인지 고민하고
혼자서 해결할 수 있는지 아니면 남들의 도움(혹은 협의)가 필요한지도 고민해보세요!

songsunkook committed:
refactor: Admin OwnerResponse 구조 개선

1. OwnerResponse가 UserResponse를 상속하도록 구조 변경
2. OwnerResponse에는 UserResponse 외에 사업자등록번호와 첨부파일 및 상점의 id들만을 가지도록 축소

Invidam commented on src/main/java/koreatech/in/mapstruct/admin/user/OwnerConverter.java:
```diff
@@ -1,13 +1,42 @@
+    @Named("convertAttachments")
+    default List<Integer> convertAttachmentsId(List<koreatech.in.domain.User.owner.OwnerAttachment> attachmentsId) {
+        return attachmentsId.stream()
+            .map(attachment -> attachment.getId()).
```
```suggestion
            .map(OwnerAttachment::getId).
```

Invidam commented on src/main/java/koreatech/in/mapstruct/admin/user/OwnerConverter.java:
```diff
@@ -1,13 +1,42 @@
+    OwnerResponse toOwnerResponse(Owner owner);
+
+    @Named("convertAttachments")
+    default List<Integer> convertAttachmentsId(List<koreatech.in.domain.User.owner.OwnerAttachment> attachmentsId) {
```
`attachmentsId` 를 사용한 이유가 있을까요?

songsunkook replied:
`@Mapping`에 대한 지식이 아직 부족했어서 사용한 것 같습니다. 수정하겠습니다!

songsunkook replied:
id 주입 방법을 변경하다보니 이게 더 적절한 것 같아서 다시 남겨두었습니다!

Invidam commented on src/main/java/koreatech/in/mapstruct/admin/user/OwnerConverter.java:
```diff
@@ -1,13 +1,42 @@
+    }
+
+    @Named("convertShops")
+    default List<Integer> convertShops(List<koreatech.in.domain.Shop.Shop> shopsId) {
```
`Shop` 대신 `koreatech.in.domain.Shop.Shop`을 사용한 이유가 있을까요?

songsunkook replied:
수정하겠습니다!

Invidam commented on src/main/java/koreatech/in/service/admin/AdminUserServiceImpl.java:
```diff
@@ -396,6 +401,19 @@ public NewOwnersResponse getNewOwners(NewOwnersCondition condition) {
     }
 
+    @Override
+    @Transactional(readOnly = true)
```
👍

Invidam reviewed (approved):
저번에 남겼던 요구사항들이 잘 반영됐네요!

이번이 진짜 마지막 코드 리뷰일 것 같아요
조금 남았으니 화이팅! ㅎ

o(*^▽^*)┛

Invidam commented on src/main/resources/mapper/admin/AdminUserMapper.xml:
```diff
@@ -69,6 +73,12 @@
             ON
                 `t1`.`id` = `t4`.`user_id`
                 AND `t4`.`is_deleted` = 0
+            LEFT JOIN `koin`.`shops` `t5`
+            ON
+                `t1`.id = `t5`.owner_id AND `t5`.is_deleted = FALSE
+            LEFT JOIN `koin`.`owner_attachments` `t6`
+            ON
+                `t1`.id = `t6`.owner_id AND `t6`.is_deleted = FALSE
```
55, 57줄에는 각각 attachments, shops 순서로 되어있는데,
여기는 shops, attachments순으로 되어있네요

응답이 뒤바뀌어서 나옵니다.

PR올리기 전에 꼼꼼히 테스트 부탁해요

songsunkook replied:
헉... 죄송합니다..!! 알려주셔서 감사합니다. 바로 수정하겠습니다!

Invidam commented on src/main/resources/mapper/admin/AdminUserMapper.xml:
```diff
@@ -69,6 +73,12 @@
             ON
                 `t1`.`id` = `t4`.`user_id`
                 AND `t4`.`is_deleted` = 0
+            LEFT JOIN `koin`.`shops` `t5`
+            ON
+                `t1`.id = `t5`.owner_id AND `t5`.is_deleted = FALSE
+            LEFT JOIN `koin`.`owner_attachments` `t6`
+            ON
+                `t1`.id = `t6`.owner_id AND `t6`.is_deleted = FALSE
```
`AdminMapper.getUserById`에서 `shops`,  `attachments` 에 대한 조인문을 추가하면

`AdminMapper.getUserById` 을 활용하고 있는 많은 곳에서 불필요한 조인문들이 반복적으로 실행될 것 같아, 개선이 필요해보입니다!
(Owner 뿐만 아니라 Student, User, 로그인 등)

songsunkook replied:
수정했습니다....!!

Invidam reviewed (changes requested):
실행해보며 발견한 문제들 있어 리뷰 남깁니다
반영부탁해요~

songsunkook committed:
refactor: join문을 별도의 쿼리로 변경
